### ISSUE TITLE
Please add full screen option for VNC

### ISSUE THREAD

**Issue by gaionim** [cockpit-project/cockpit-machines#680]
As said from mzamot in [#8392 ](https://github.com/cockpit-project/cockpit/issues/8392#issuecomment-566170769)

Having a full screen button or the option of opening the VNC in a different tab would be really nice

I'll try to do it by myself, but I'm new to react/js :-(

**Comment by skobyda** [comment]
I believe it could be achieved. Our vnc display is basically a single react component <VncConsole />
We can either make the VNC display component to stretch to the 100% height of the browser.
Or we truly make it fullscreen, there seem to be some npm libraries that can lock the whole component to the fullscreen. 

I'm now not sure if it should be a replacement for our already existing "Expand" option.
Or maybe just an alternative, offerend next to "Expand", something like:
![Screenshot from 2022-04-21 14-43-19](https://user-images.githubusercontent.com/42733240/164460524-c3a58efe-1afd-4d45-87a7-6db42be65c0c.png)


@garrett WDYT?

**Comment by skobyda** [comment]
It would also have some fitting icon next to "Fullscreen" just like expand option has an icon.

**Comment by garrett** [comment]
You can already expand VNC in Cockpit-Machines.

![Screenshot 2022-04-21 at 14-25-01 Virtual Machines - garrett@Bolt](https://user-images.githubusercontent.com/10246/164457668-3e18becb-8b28-4c70-bc6c-fc15f9dd5319.png)

Then click expand

![image](https://user-images.githubusercontent.com/10246/164457810-baeab886-cd0d-4c28-98b5-0aa5bdd49e4d.png)


And you'll have a larger view.

![Screenshot 2022-04-21 at 14-26-07 Virtual Machines - garrett@Bolt](https://user-images.githubusercontent.com/10246/164457686-277143a1-de73-49b2-b00f-c0a108efb332.png)

However, the bar is still at the top and the navigation is still there on the side.

I guess you'd prefer a pop-out (with minimal Cockpit UI) or fullscreen option instead? (I'm not sure doing true fullscreen is possible.)

**Comment by garrett** [comment]
I think we'd want to make a minimal version or fullscreen (if possible) replace the current expand functionality.

**Comment by gaionim** [comment]
My naif idea was to add the alternative to open in new tab/window only the single component and delagate to the browser the fullscreen

**Comment by garrett** [comment]
@gaionim: Yeah, that'd be nice. That's what I meant as a popout option. It would be an almost-empty new tab (or window) with just the terminal. Then you could then maximize that or even hit F11 to fullscreen the browser window. Or we could just handle it as a sub page with a very minimal link at the top to go back (without all the rest of the Cockpit UI).

Using fullscreen support directly is special, and I'm not sure if that's possible in a reasonable way (as, IIRC, hitting ESC will kick you out of fullscreen, breaking a lot of apps, especially vim). https://developer.mozilla.org/en-US/docs/Web/API/Fullscreen_API

I guess the best option would be to have an extremely minimal page with just a back link at the top and the terminal sized to the rest. And if you middle click it'd open in a tab. Otherwise, use the breadcrumb link at the top.

**Comment by garrett** [comment]
It'd basically look like this if we stripped out everything not 100% relevant to the terminal:

![image](https://user-images.githubusercontent.com/10246/165333473-b604e527-aa52-4eb2-816b-caafcfe1d4f4.png)

(This is just a modification of one of the above screenshots.)

I do have a different mockup @ https://github.com/cockpit-project/cockpit-machines/issues/553#issuecomment-1061581973 for when the terminal is embedded, regarding how to handle setting up ports and configuration. The other issue could affect how this works too. Perhaps we want the widget to basically zoom on page to take up the entire page instead?

![mockup](https://user-images.githubusercontent.com/10246/157207848-27569e7d-5709-414d-9ac9-2c04945534b8.png)

And it could then possibly resize over everything else, where it stretches like this:

![expanded](https://user-images.githubusercontent.com/10246/165335161-a7c25a25-e804-4260-adf3-f4dd78f84744.png)
(Expand would be collapse with a different icon.)

**Comment by gaionim** [comment]
> Using fullscreen support directly is special, and I'm not sure if that's possible in a reasonable way (as, IIRC, hitting ESC will kick you out of fullscreen, breaking a lot of apps, especially vim). https://developer.mozilla.org/en-US/docs/Web/API/Fullscreen_API


I think that full screen is useful when working with gui, when the click of the mouse  trigg the window/frame scroll;
for text program ( like vim ) I ( whe) can work nice with  "expand" ;-)

**Comment by garrett** [comment]
Fullscreen means different things. You can hit F11 to maximize your browser and we can make some changes so there's not as much Cockpit around VNC. That's what I'm talking about being possible to implement.

The **fullscreen API** is almost assuredly _not_ what you (and most other people) actually want, as if you hit the ESC key in vim while using the embedded VNC viewer, it will stop fullscreen mode and won't be passed to vim... which makes vim kind of impossible to use, unless you know the alternate escape key sequence, `Ctrl`-`[`... but most people use ESC.

What you _probably_ want is a page where VNC takes up almost all of it instead. (Almost all, so we have very minimal UI to get back to Cockpit.) And then you can either just have a larger VNC area than right now and you can maximize it or even fullscreen your browser window (but not the VNC part itself).

Summary: I agree; maximizing VNC is a great thing Cockpit Machines should do. (But we shouldn't use the fullscreen API, which is meant for games and would pose some issues for us.)

**Comment by johnjtran** [comment]
Great discussion.  How about a pop-up/detachable window? so `Expand` and `Detach`?

**Comment by shodanshok** [comment]
> Great discussion. How about a pop-up/detachable window? so `Expand` and `Detach`?

Yeah, it would be nice to have a detachable windows, maybe in addition to the existing (or, even better, the proposed "trimmed") expand option.

**Comment by Betonhaus** [comment]
is it possible to resize the screen to fit in the window? I can never get the expand view to fully fit so I always have to pan up and down to navigate
![image](https://github.com/user-attachments/assets/74631d70-6990-41a6-a770-9195edd613f3)

**Comment by tgwaste** [comment]
Don't hold your breath. For some reason they just don't want to improve this experience. This is the worst aspect of cockpit imo. A resizable pop-out would make this whole thing 1000x better.

**Comment by Betonhaus** [comment]
> Don't hold your breath. For some reason they just don't want to improve this experience. This is the worst aspect of cockpit imo. A resizable pop-out would make this whole thing 1000x better.

As it being a Fedora affiliated project that's a given, but I'm sure eventually they can be convinced that this is a valid use case. I have Remote Desktop set up but when not on the local network or doing configurations it's nice to be able to pull it into a proper view.

**Comment by mac2net** [comment]
I do not have knowledge of internal discussions of priorities or how resources get allocated, but after observing Cockpit development for a couple of years, it is my opinion that the functions of Cockpit have improved in most areas of the app. Frankly the VNC option in Cockpit is really more one of convenience – which it is – then for day to day GUI management. I setup XFCE on the metal and VMs - a light weight DE – and VNC into that when the need arises – it could be hourly or only every two weeks. There are certain GUI apps – while still avoiding FlatPacks – that add value and – make it worth it for the user to invest the extra time to install it but is not necessarily a priority of the Cockpit project as a whole.

**Comment by martinpitt** [comment]
> they just don't want to improve this experience

It's not willingness or malice, we just haven't had a proper cockpit-machines developer for a long time now, so we can only do life support and a little maintenance these days. :cry: If someone has some time and dedication to improve this, we'll gladly help/review/give guidance etc.

**Comment by Betonhaus** [comment]
> > they just don't want to improve this experience
> 
> It's not willingness or malice, we just haven't had a proper cockpit-machines developer for a long time now, so we can only do life support and a little maintenance these days. 😢 If someone has some time and dedication to improve this, we'll gladly help/review/give guidance etc.

There were a few things that I felt could improve the experience, but I'll concede that it's largely functional as it is.
I will note the issues that I had when setting up a Windows VM, though I don't know how feasible they are to deal with
1. I had a seperate SSD that I wanted to dedicate to the VM. AppArmor blocks access to anything that isn't in the default libvirt image directory and it isn't made clear how to configure a data pool on a different drive to be whitelisted, and I ended up having to disable the relevant security feature as that what people with the same issues ended up doing - there didn't seem to be anything in documentation about this
2. When adding a shared folder to a Windows guest I had to find a guide to  find and install relevant drivers. The guides are easy to find, however.
3. The Windows VM had a severe performance hit, which required editing the VMs .XML file directly to disable several scheduling systems and pinning specific CPU cores as [per this guide](https://www.reddit.com/r/ChromeOSFlex/comments/ucno4b/comment/i6cviv8/?utm_source=share&utm_medium=mweb3x&utm_name=mweb3xcss&utm_term=1&utm_content=share_button), something that I feel should be partly the default configuration for windows guests, and partly exposed in the settings page for the VM
I concede that if the project is understaffed there's not much I can do, the software seems sufficiently stable and I'd have to wait a while for fixes anyways as I'm using the Debian repository for cockpit

**Comment by garrett** [comment]
We definitely want to improve the VM part of Cockpit machines, and I have even worked on this feature a few times over the years, making mockups and spec'ing out how it should work, and have pushed for this to be improved more than a few times. However, it requires someone dedicated to the task to implement it.

**Comment by Venefilyn** [comment]
Doing a popup seems quite difficult tbh due to restrictions in browsers, and then we'd have to think about other areas like the Cockpit Client etc.

Maybe invoking fullscreen would be best where we hide all the Cockit related things that isn't VM specific. So top bar, sidebar, etc.

Then we could make sure the top bar is always visible so one can actually exit fullscreen easily. It might also mean ensuring Escape won't trickle into the browser to escape fullscreen too -- but not sure how that would work yet

**Comment by mvollmer** [comment]
Everybody, please have a look at this demo: https://www.youtube.com/watch?v=FsBE4eQo0As

(Watch it in fullscreen! :-)

We are ready and eager to implement some fullscreen goodness, but need feedback on how it should look exactly.

Some questions we are struggling with:

 - Is it OK to eat the Escape key while in full-monitor-mode? We could do the same thing as Youtube does to go fullscreen, but the browser would intercept the Escape key. Would that be a problem? (We can add it into the "Send keys" menu, but vim users will likely go mad.)

 - In addition to a full-monitor-like-youtube mode, we would also provide a "full-browser-window" mode (like the Youtube Cinema mode). This would replace the existing "Expand" functionality. The question: Should we keep it as a "expand/collapse" toggle operation, or should we have a "Open in new window" button.

 - How would you feel if instead of the "full-browser-window" mode we would allow the navigation sidebar to be toggled in and out? So clicking "Expand" would fill the content area like it does now, and if you need a bigger console still, you have to toggle away the sidebar explicitly.

Thanks!

**Comment by martinpitt** [comment]
> [fullscreen] We can add it into the "Send keys" menu, but vim users will likely go mad.

No, we can't, these are mutually exclusive. In fullscreen you don't see these controls, and in window mode pressing Esc works just fine. If you need Esc, it's F11 or normal window mode.

**Comment by mvollmer** [comment]
> In fullscreen you don't see these controls,

My current assumption is that we keep these controls accessible always.

**Comment by martinpitt** [comment]
@mvollmer how do you want to do that? Is there some CSS magic to overlay the "send special key" on top of the full-screen noVNC element? That would be rather annoying, no?

**Comment by mvollmer** [comment]
> [@mvollmer](https://github.com/mvollmer) how do you want to do that? Is there some CSS magic to overlay the "send special key" on top of the full-screen noVNC element? That would be rather annoying, no?

Initially like in the demo: there is a thin(?) strip of buttons above the noVNC element.

**Comment by Venefilyn** [comment]
> [@mvollmer](https://github.com/mvollmer) how do you want to do that? Is there some CSS magic to overlay the "send special key" on top of the full-screen noVNC element? That would be rather annoying, no?

Just to chime in on what you can do. It is possible depending on how you fullscreen the elements, [`requestFullscreen()`](https://developer.mozilla.org/en-US/docs/Web/API/Element/requestFullscreen) can be called on specific elements which will make that element and all its children part of the fullscreen. In #2241 I did `document.getElementById("virtual-machines-page")?.requestFullscreen();` which would get all the controls at the top.

If you want a real life example fullscreen a video on YouTube and scroll and you see they have UI there and not just the video
